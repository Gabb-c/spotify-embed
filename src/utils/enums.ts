/** List of available embed spotify URL's */
export enum SpotifyEmbedUrl {
  TRACK = 'https://open.spotify.com/embed/track/',
  ALBUM = 'https://open.spotify.com/embed/album/',
  PLAYLIST = 'https://open.spotify.com/embed/playlist/',
  ARTIST = 'https://open.spotify.com/embed/artist/',
  SHOW = 'https://open.spotify.com/embed/show/',
  EPISODE = 'https://open.spotify.com/embed/episode/',
}

export enum SpotifyEmbedTypes {
  TRACK = 'track',
  ALBUM = 'album',
  PLAYLIST = 'playlist',
  ARTIST = 'artist',
  SHOW = 'show',
  EPISODE = 'episode',
}

export enum Default {
  WIDTH = 300,
  HEIGHT = 380,
}

export enum Compact {
  WIDTH = 300,
  HEIGHT = 80,
}

export enum Themes {
  LIGHT = '1',
  DARK = '0',
}
