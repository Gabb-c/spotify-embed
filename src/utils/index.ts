export * from './enums';
export * from './parser';
