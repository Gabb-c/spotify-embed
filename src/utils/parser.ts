import { SpotifyEmbedUrl, SpotifyEmbedTypes, Themes } from './enums';

/**
 * @param embedType - The type of the embed
 * @param spotifyId - The Spotify resource ID
 * @returns The formated embed url
 */
const buildUrl = (embedType: string, spotifyId: string, theme?: 'light' | 'dark'): string => {
  const chosenTheme = theme === 'dark' ? Themes.DARK : Themes.LIGHT;
  switch (embedType) {
    case SpotifyEmbedTypes.ALBUM: {
      return SpotifyEmbedUrl.ALBUM + spotifyId + `?theme=${chosenTheme}`;
    }
    case SpotifyEmbedTypes.ARTIST: {
      return SpotifyEmbedUrl.ARTIST + spotifyId + `?theme=${chosenTheme}`;
    }
    case SpotifyEmbedTypes.EPISODE: {
      return SpotifyEmbedUrl.EPISODE + spotifyId + `?theme=${chosenTheme}`;
    }
    case SpotifyEmbedTypes.PLAYLIST: {
      return SpotifyEmbedUrl.PLAYLIST + spotifyId + `?theme=${chosenTheme}`;
    }
    case SpotifyEmbedTypes.SHOW: {
      return SpotifyEmbedUrl.SHOW + spotifyId + `?theme=${chosenTheme}`;
    }
    case SpotifyEmbedTypes.TRACK: {
      return SpotifyEmbedUrl.TRACK + spotifyId + `?theme=${chosenTheme}`;
    }
    default: {
      return '';
    }
  }
};

/**
 * @param url - The Spotify resource URL
 * @param theme - The embed theme
 * @returns The formated embed URL
 */
export const urlParser = (url: string, theme?: 'light' | 'dark'): string => {
  if (url.length === 0 || !url.startsWith('https://open.spotify.com/')) return '';
  const parsed = url.split('/');

  return buildUrl(parsed[3], parsed[4]?.split('?')[0], theme);
};

/**
 * @param uri - The Spotify resource URI
 * @param theme - The embed theme
 * @returns The formated embed url
 */
export const uriParser = (uri: string, theme?: 'light' | 'dark'): string => {
  if (uri.length === 0 || !uri.startsWith('spotify')) return '';

  const parsed = uri.split(':');

  return buildUrl(parsed[1], parsed[2], theme);
};
