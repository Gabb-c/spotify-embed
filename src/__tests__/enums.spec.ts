import { Compact, Default, SpotifyEmbedTypes, SpotifyEmbedUrl, Themes } from '../utils/enums';

describe('Sizes', () => {
  it('Compact enum should match snapshot', () => {
    expect(Compact).toMatchSnapshot();
  });

  it('Default enum should match snapshot', () => {
    expect(Default).toMatchSnapshot();
  });
});

describe("Embed Types and URL's", () => {
  it('SpotifyEmbedTypes enum should match snapshot', () => {
    expect(SpotifyEmbedTypes).toMatchSnapshot();
  });

  it('SpotifyEmbedUrl enum should match snapshot', () => {
    expect(SpotifyEmbedUrl).toMatchSnapshot();
  });
});

describe('Embed Themes', () => {
  it('SpotifyEmbedTypes enum should match snapshot', () => {
    expect(Themes).toMatchSnapshot();
  });
});
