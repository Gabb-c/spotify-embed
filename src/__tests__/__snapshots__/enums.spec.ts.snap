// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Embed Themes > SpotifyEmbedTypes enum should match snapshot 1`] = `
{
  "DARK": "0",
  "LIGHT": "1",
}
`;

exports[`Embed Types and URL's > SpotifyEmbedTypes enum should match snapshot 1`] = `
{
  "ALBUM": "album",
  "ARTIST": "artist",
  "EPISODE": "episode",
  "PLAYLIST": "playlist",
  "SHOW": "show",
  "TRACK": "track",
}
`;

exports[`Embed Types and URL's > SpotifyEmbedUrl enum should match snapshot 1`] = `
{
  "ALBUM": "https://open.spotify.com/embed/album/",
  "ARTIST": "https://open.spotify.com/embed/artist/",
  "EPISODE": "https://open.spotify.com/embed/episode/",
  "PLAYLIST": "https://open.spotify.com/embed/playlist/",
  "SHOW": "https://open.spotify.com/embed/show/",
  "TRACK": "https://open.spotify.com/embed/track/",
}
`;

exports[`Sizes > Compact enum should match snapshot 1`] = `
{
  "300": "WIDTH",
  "80": "HEIGHT",
  "HEIGHT": 80,
  "WIDTH": 300,
}
`;

exports[`Sizes > Default enum should match snapshot 1`] = `
{
  "300": "WIDTH",
  "380": "HEIGHT",
  "HEIGHT": 380,
  "WIDTH": 300,
}
`;
